platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingArtifactVariantSelector.java

```diff
@@ -145,28 +145,14 @@ private static List<TransformedVariant> tryDisambiguate(
         // Find any other candidate which does not match with the last candidate.
         for (int i = 0; i < matches.size() - 1; i++) {
             TransformedVariant current = matches.get(i);
-            if (candidatesDifferent(matcher, current, last)) {
+            if (!matcher.weaklyMatches(current.getAttributes(), last.getAttributes())) {
```

Comment:
The javadoc explains exactly what this method does. It could probably do a better job explaining _why_ it is doing it. Most of that explanation you can find in my prior comment.

I can update the javadoc a bit but I don't think we need to update the algorithm here at all. 

Part of the trouble making this a stream is that it tracks mutable state (`differentTransforms`) outside of the loop. Streams arent good at doing that without using some kind of `reduce` operation, and I would bet that such a change would be _harder_ to comprehend over a simple loop and an if statement. 

Its funny you call this really bloated. 

You should see what it used to look like before I touched it:
https://github.com/gradle/gradle/blob/6a6618e4cf86c690972c7d075e24187a8df8c490/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingVariantSelector.java#L154-L228


[commit](https://github.com/gradle/gradle/commit/3bc0246c02fe126f5f5f03dfec13ed2f12f9202f#diff-8a6557ba4bdf62bad6a3668df773bcf6bd55795553a5e52ec777b5a6d5df08ee)

Reply:
The javadoc used to define "tryDisambiguate" as "Attempt to disambiguate", which was completely opaque as what disambiguation was.  The change is much better.